rfcs/20230118-f16-accumulation/README.md

```diff
@@ -0,0 +1,125 @@
+and many data show that good levels of accuracy can be maintained
+while using f16 accumulation datatype for most f16 models. We hence
+have been dispatching implementations with f16 accumulation on some
+platforms.
+
+Even though dispatching implementations with f16 accumulation is
+acceptable for inference, it is not for training, which requires the
+full f32 precision accumulators. Hence, as oneDNN is planning
+support for f16 training, it is important that we dispatch only
+implementation with f32 accumulation for training workloads.
+
+Note that the main user for oneDNN f16 implementations is openVINO,
+which is the requester for the implementations with f16 accumulation
+in the first place.
+
+
+# 3. Options
+
+## Using the propagation kind
+
+One natural option would be to rely on propagation kind.  If the
+propagation kind, is `forward_inference`, we allow dispatching of
+implementations with f16 accumulation, otherwise we don't.  This works
+for most primitives except for Matmul, which does not have a
+propagation kind. Hence this option does not cover the whole library.
+
+## Using the `ONEDNN_ENABLE_WORKLOAD` build option
+
+oneDNN currently has a `ONEDNN_ENABLE_WORKLOAD` build option that can
+be set to `INFERENCE` or `TRAINING` (default). This option is
+currently used to disable dispatching of some primitive
+implementations based on propagation_kind. We could rely on this
+mecanism to guard dispatching of implementations using f16
+accumulation with the two following caveats:
+- this would increase validation costs as we would need to validate both builds
+- it adds development burden. Currently we don't know any information
+  on the accumulation datatype for each implementations. Properly
+  guarding implementation dispatch would either need to be scattered
+  in implementation using f16 accumulation (which is brittle), or
+  would require to add internal APIs to query implementations for their
+  accumulation datatype.
+
+## A new `accumulation_datatype` attribute (recommended)
+
+The last option is to rely on a new primitive attribute to get a hint
+on minimal accuracy required in accumulators. There are multiple
+options for what values it should accept.
+
+For all those options, note that implementations would still be able
+to use a different datatype internally if the result is unchanged and
+it is faster (e.g. using `f32` intermediate accumulators when
+accumulation datatype is `f16`, or in some `s8/u8` computations when
+it yields the same result).
+
+### Use a datatype (recommended).
+
+In those cases, the user will have to explicitely pass `f32`, `f16`,
+`s32`, ...  When the user does not set the attribute, we will use a
+default value of `undef`, which would allow all implementation to be
+dispatched, including those with `f16` accumulation.
+
+Pros:
+- no new datatype introduced
+- the user explicitly passes the datatype
+
+Cons:
+- not fully clear from API that it is a hint
```

Comment:
how does the user know if the hint is applied?
Not entirely sure if we can rely on impl/isa string, as some primitives use highest supported impl/isa. 
Maybe additional field in verbose string?

Reply:
This is a good question, and we face a similar dilemma than fpmath mode.

We will have to print something in verbose, it will either be what was used to create the primitive (useful for reproducer), or what is effectively used by the primitive (useful for user information).
I would align with fpmath mode behavior here and print the value of the attribute as passed, not as interpreted by the impl.

To print the accumulation datatype that is effectively used, we can use new verbose extension that can print this kind of extra information.

Reply:
(Just FYI) Currently it is possible to know if fpmath attribute is used if the implementation string has `amx` in it... although it is only applicable for CPU.